### PR TITLE
Make sure to clear tag word when upgrading a struct list.

### LINKED ISSUE #419: tag not cleared in struct list upgrade
Consider the following schema and program:

```capnp
# upgrade.capnp
@0xca1a6fd5ed3b4bb1;

struct RootStruct {
   field @0 :AnyPointer;
}

struct OldVersion {
   old0 @0 :UInt32;
   old1 @1 :UInt32;
}

struct NewVersion {
   old0 @0 :UInt32;
   old1 @1 :UInt32;
   new2 @2 :UInt64;
}
```

```cpp
// upgrade.c++

#include "upgrade.capnp.h"
#include <capnp/message.h>
#include <capnp/serialize.h>
#include <capnp/serialize.h>

int main() {
  ::capnp::MallocMessageBuilder message;
  auto root = message.initRoot<RootStruct>();
  auto oldList = root.getField().initAs<capnp::List<OldVersion>>(1);
  oldList[0].setOld0(0xab);
  oldList[0].setOld1(0xcd);

  // Reallocate the list with enough space for the new version of the struct.
  auto newList = root.getField().getAs<capnp::List<NewVersion>>();

  capnp::writeMessageToFd(1, message);
}
```

I expect this to write a message that looks like this:

```
0000 0000 0700 0000 // one segment with 7 words
0000 0000 0000 0100 // struct pointer, pointer section has one pointer
0900 0000 1700 0000 // list pointer, offset 2, inline composite, 2 words long
0000 0000 0000 0000 // [zeroed old list tag]
0000 0000 0000 0000 // [zeroed old list body]
0400 0000 0200 0000 // list tag. 1 element. data section has 2 words.
ab00 0000 cd00 0000 // first element, first word
0000 0000 0000 0000 // first element, second word
```

Instead, it writes this message:

```
0000 0000 0700 0000 // one segment with 7 words
0000 0000 0000 0100 // struct pointer, pointer section has one pointer
0900 0000 1700 0000 // list pointer, offset 2, inline composite, 2 words long
0400 0000 0100 0000 // old list tag! (BUG)
0000 0000 0000 0000 // [zeroed old list body]
0400 0000 0200 0000 // list tag. 1 element. data section has 2 words.
ab00 0000 cd00 0000 // first element, first word
0000 0000 0000 0000 // first element, second word
```

### PR DESCRIPTION
**Pull request by dwrensha**
Fixes #419.

### CODE REVIEW

File: c++/src/capnp/layout.c++ [hunk 1]
```diff
@@ -1360,7 +1360,8 @@ struct WireHelpers {
       }
 
       // Zero out old location.  See explanation in getWritableStructPointer().
-      memset(oldPtr, 0, oldStep * elementCount * BYTES_PER_WORD / BYTES);
+      // Make sure to include the tag word.
+      memset(oldPtr - 1, 0, (1 + oldStep * elementCount) * BYTES_PER_WORD / BYTES);
```
Comment (kentonv) [inline c++/src/capnp/layout.c++]: I think this will fail unit analysis. The `1` needs to be `1 * WORDS`.

Comment (dwrensha) [inline c++/src/capnp/layout.c++]: How do I run the unit analysis?

Comment (kentonv) [inline c++/src/capnp/layout.c++]: Build with `-DCAPNP_DEBUG_TYPES`.

Comment (kentonv) [inline c++/src/capnp/layout.c++]: Ah, yes, that is more correct!

**Pull request merged**